test/fiber/scheduler.rb

```diff
@@ -136,13 +135,51 @@ def wait_any(io, events, duration)
   end
 
   def wait_for_single_fd(fd, events, duration)
-    wait_any(
-      for_fd(fd),
-      events,
-      duration
-    )
+    wait_any(for_fd(fd), events, duration)
   end
 
+  # called from Mutex#lock
+  def wait_mutex(mutex)
+    p [:wait_mutex, mutex]
+    # @waiting_on[Fiber.current] = mutex
+
+    # Guarantee: Fiber.current is a Fiber of current Thread and is a Fiber of this scheduler
+    @waiting_mutex[mutex] << Fiber.current
+    Fiber.yield
+    true
+  end
+
+  # called from Mutex#unlock, possibly from another Fiber on the same thread, or possibly from some other thread
+  def notify_mutex(mutex)
+    p [:notify_mutex, mutex]
+    # p @waiting_mutex
+    q = @waiting_mutex[mutex]
+    fiber = q.shift
+    @waiting_mutex.delete(mutex) if q.empty?
+    raise unless fiber
+
+    if Thread.current == @thread
+      # but what if fiber died in the meanwhile?
+      # Is it possible if same thread? Yes, due to Fiber#raise
+      # And what if it somehow went further in the meanwhile, due to e.g. Fiber#raise, we'll resume another point?
+      if fiber.alive?
+        p [:resume, fiber]
+        fiber.resume
+      end
+    else
+      # scheduler = Thread.current.scheduler
+      # send -> { notify_mutex(mutex) } to scheduler Queue + control pipe write to wake up?
+      # (could be abstracted as #external_notification(&block) ?)
+      # or maybe shouldn't come at all in #notify_mutex ? => seems simpler
+      # Best seems to remember scheduler+Mutex instance (+Fiber?) in `waitq` for non-blocking waiters,
+      # so we know exactly how to wake up a given waiter Fiber.
```

Comment:
This is about how should we deal with a mix of blocking and non-blocking Fibers?
I guess a simple way is to remember the scheduler on contended Mutex#lock in the wait queue along with the Fiber & Thread. That way when unlocking we remember how to wake up that particular waiter.

Reply:
It's entirely possible a scheduler comes in and out of existence between lock/unlock steps. So, I'm not sure we can rely on the scheduler actually doing anything. The notification should be a best effort. I think I need to see the actual implementation, but we should be able to essentially use `external_notification`/`notify` as a best effort "the mutex is available now, try again".

The wait list in the scheduler is most efficient if it's a list of fibers to call resume on.

So, thinking about it from this point of view, maybe makes sense:

``` ruby
def notify(fiber, urgent = false)
```

This would allow you to simply call

```
Thread.current.scheduler.notify(mutex->fiber, true)
```